### PR TITLE
feat(KongPlugin): generate K8s Events for misconfiguration

### PR DESCRIPTION
**Pull request by programmer04**
**What this PR does / why we need it**:

One of the cases mentioned in the original issue https://github.com/Kong/kubernetes-ingress-controller/issues/6540:

`a Plugin on which an HTTPRoute depends`

can be extrapolated to any plugin attachment referring to a nonexisting plugin leads to a logs message, but no K8s Event is sent.

Status for `HTTPRoute` that references non-existing plugin is

```yml
...
  status:
    parents:
    - conditions:
      - lastTransitionTime: "2024-12-12T16:06:22Z"
        message: ""
        observedGeneration: 1
        reason: Accepted
        status: "True"
        type: Accepted
      - lastTransitionTime: "2024-12-12T16:06:22Z"
        message: ""
        observedGeneration: 1
        reason: ResolvedRefs
        status: "True"
        type: ResolvedRefs
      - lastTransitionTime: "2024-12-12T16:06:22Z"
        message: ""
        observedGeneration: 1
        reason: Unknown
        status: Unknown
        type: Programmed
      controllerName: konghq.com/kic-gateway-controller
      parentRef:
        group: gateway.networking.k8s.io
        kind: Gateway
        name: kong
        namespace: default
```

and respective K8s Events are presented

```log
default              2m51s       Warning   KongConfigurationTranslationFailed   httproute/httproute-testing                           referenced KongPlugin or KongClusterPlugin "my-amazing" does not exist
default              2m51s       Warning   KongConfigurationTranslationFailed   httproute/httproute-testing                           referenced KongPlugin or KongClusterPlugin "second" does not exist
```

The mechanism for setting appropriate status is implemented only for generated controllers with the helper [EnsureProgrammedCondition](https://github.com/Kong/kubernetes-ingress-controller/blob/67c8a9596a3ec1c8d1d87a269c12eb491dbf25da/internal/controllers/utils/conditions.go#L36-L107), for make them robust, I'm suggesting creating a dedicated issue.

There are still logs that should be events and statuses related to plugins like in [this helper](https://github.com/Kong/kubernetes-ingress-controller/blob/67c8a9596a3ec1c8d1d87a269c12eb491dbf25da/internal/dataplane/kongstate/kongstate.go#L392-L429) for cross-namespace access, but it can be handled in a separate PR or issue .

**Which issue this PR fixes**:

https://github.com/Kong/kubernetes-ingress-controller/issues/6540

**Special notes for your reviewer**:

I wonder whether performance and/or memory consumption will be significantly damaged...

I'm going to create an issue for the enhancement of statuses and removing those other logs for missing/misconfigured Plugins

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/util/relations.go [hunk 1]
```diff
@@ -1,11 +1,29 @@
 package util
 
+import "sigs.k8s.io/controller-runtime/pkg/client"
+
+// FR represents a foreign relation.
+type FR struct {
+	Identifier string
+	// Referer represents K8s object that references a Plugin.
+	Referer client.Object
+}
```
Comment (pmalek) [inline internal/util/relations.go:10]: I'm OK leaving this in in this PR but IMHO we should move all this code out to a dedicated package at some point where all foreign relations code is put. WDYT?

File: internal/dataplane/kongstate/plugin.go [hunk 2]
```diff
@@ -18,33 +18,34 @@ import (
 )
 
 // getKongPluginOrKongClusterPlugin fetches a KongPlugin or KongClusterPlugin (as fallback) from the store.
-// If both are not found, an error is returned.
+// If both don't exist pluginFound is set to false.
```
Comment (pmalek) [inline internal/dataplane/kongstate/plugin.go:21]: WDYT about not adding the boolean return value but instead return `store.NotFoundError` when neither plugin nor cluster plugin are found and then handle that error type at the call site checking the type of the error?

**Review by pmalek — COMMENTED** [review]
I like this change 👍 

> I wonder whether performance and/or memory consumption will be significantly damaged...

I wouldn't be too concerned about this. The code looks sensible and we're making the change to provide user facing change: create Events instead of logging.

> I'm going to create an issue for the enhancement of statuses and removing those other logs for missing/misconfigured Plugins

👍

File: internal/util/relations.go (hunk 1 above)
Comment (programmer04) [reply internal/util/relations.go:10]: This is a sensible idea to do in a separate PR

File: internal/dataplane/kongstate/plugin.go (hunk 2 above)
Comment (programmer04) [reply internal/dataplane/kongstate/plugin.go:21]: I followed convention with maps that you have a boolean to indicate whether a value exists or not. Now it makes an error reserved only for something that feels to me like a "real" error and does not require from a caller understanding of dedicated error for checking existence. 

I always feel a little bit uncomfortable when errors are used to report something that may be a result of normal usage. In a language with proper discriminated unions (F# naming), variants/tagged unions (OCaml naming), or enums in Rust, I can model it with one return value that can be `KP | KCP | None | Err` and deconstruct it with exhaustive pattern matching. Unfortunately, Go does not offer it, so for me the most sensible/readable was to have a dedicated boolean to indicate whether something was found or not without errors

So to sum up, IMHO it can be either way in Go :D

Comment (pmalek) [reply internal/dataplane/kongstate/plugin.go:21]: Fair enough 👍 Let's leave it as is.

**Review by pmalek — APPROVED** [review]
